client/ladder/action-creators.ts

```diff
@@ -3,33 +3,52 @@ import { MatchmakingType } from '../../common/matchmaking'
 import { apiUrl, urlPath } from '../../common/urls'
 import { ThunkAction } from '../dispatch-registry'
 import { push } from '../navigation/routing'
+import { abortableThunk, RequestHandlingSpec } from '../network/abortable-thunk'
 import { fetchJson } from '../network/fetch'
 
 const LADDER_RANKINGS_CACHE_TIME_MS = 60 * 1000
 
-export function getRankings(matchmakingType: MatchmakingType): ThunkAction {
-  return (dispatch, getState) => {
+const lastFetchTimeByMatchmakingType = new Map<MatchmakingType, Date>()
+const lastSearchQueryByMatchmakingType = new Map<MatchmakingType, string>()
+
+export function getRankings(
+  matchmakingType: MatchmakingType,
+  searchQuery: string,
+  spec: RequestHandlingSpec<void>,
+): ThunkAction {
+  return abortableThunk(spec, async dispatch => {
     const fetchTime = new Date()
-    const {
-      ladder: { typeToRankings },
-    } = getState()
-    const rankings = typeToRankings.get(matchmakingType)
-    const lastFetchTime = rankings?.fetchTime ?? 0
+    const lastFetchTime = lastFetchTimeByMatchmakingType.get(matchmakingType) ?? new Date(0)
+    const lastSearchQuery = lastSearchQueryByMatchmakingType.get(matchmakingType) ?? ''
 
     if (
-      !rankings?.lastError &&
-      Number(fetchTime) - Number(lastFetchTime) < LADDER_RANKINGS_CACHE_TIME_MS
+      Number(fetchTime) - Number(lastFetchTime) < LADDER_RANKINGS_CACHE_TIME_MS &&
+      lastSearchQuery === searchQuery
     ) {
       return
     }
 
-    dispatch({ type: '@ladder/getRankingsBegin', payload: { matchmakingType, fetchTime } })
+    lastFetchTimeByMatchmakingType.set(matchmakingType, fetchTime)
+    lastSearchQueryByMatchmakingType.set(matchmakingType, searchQuery)
+
+    const result = await fetchJson<GetRankingsResponse>(
+      apiUrl`ladder/${matchmakingType}`.concat(searchQuery ? urlPath`?q=${searchQuery}` : ''),
+      {
+        signal: spec.signal,
+      },
+    )
+
+    if (fetchTime < lastFetchTime) {
```

Comment:
So what you're trying to do here is good, but this implementation doesn't actually accomplish it. What you want is "check the current lastFetchTime value to make sure it is the same as when we started the fetch". What you have here is "check the value of the local variable we assigned at the start of the function to make sure it hasn't changed". This function is the only owner of that variable so we can guarantee it will *never* change in this code, thus this code is dead.

What you actually want is:
```ts
if (fetchTime < lastFetchTimeByMatchmakingType.get(matchmakingType))
```

I think this would also be more readable if you reversed the condition and put the dispatch inside of that, rather than using an early return:

```ts
if (fetchTime >= lastFetchTimeByMatchmakingType.get(matchmakingType)) {
  dispatch({
    type: '@ladder/getRankings',
    payload: result,
    meta: { matchmakingType },
  })
}
```